### PR TITLE
security.acme: added reload and restart lists for sytemd targets

### PR DESCRIPTION
**Pull request by qknight**
###### Motivation for this change

* added a list for systemd reload and restart targets
* using the nixos module system to set the default at the domain option (instead of using code in the `config = ` section which duplicates code and makes it not accessible from outside the acme.nix scope)

###### Things done

- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: nixos/modules/security/acme.nix [hunk 1]
```diff
@@ -20,8 +20,8 @@ let
       };
 
       domain = mkOption {
-        type = types.nullOr types.str;
-        default = null;
+        type = types.str;
```
Comment (zimbatm) [inline nixos/modules/security/acme.nix]: is it not possible to keep backward-compatibility?

File: nixos/modules/security/acme.nix [hunk 2]
```diff
@@ -52,6 +52,36 @@ let
...
+      systemd = mkOption {
+        description = "submodule example";
+        default = {};
+        type = with types; submodule {
+          options = {
+
+            reload = mkOption {
+              type = types.listOf types.str;
+              default = [];
+              example = [ "postfix.service" ];
```
Comment (zimbatm) [inline nixos/modules/security/acme.nix]: `postfix.service` is used in both examples. Which one is needed for postfix and can we find a different common example for the other-one?

File: nixos/modules/security/acme.nix [hunk 3]
```diff
@@ -193,10 +225,11 @@ in
           servicesLists = mapAttrsToList certToServices cfg.certs;
           certToServices = cert: data:
               let
-                domain = if data.domain != null then data.domain else cert;
+                restartJobs = lib.unique data.systemd.restart;
+                reloadJobs = lib.subtractLists restartJobs (lib.unique data.systemd.reload);
```
Comment (zimbatm) [inline nixos/modules/security/acme.nix]: What is the case where a restart jobs is promoted to reload only? It seems like a configuration error.

**Comment by qknight** [comment]
i have to fix a issue with the overall design first and then i will reopen the PR

**Comment by zimbatm** [comment]
sure no problem

File: nixos/modules/security/acme.nix (hunk 3 above)
Comment (qknight) [reply nixos/modules/security/acme.nix]: @zimbatm i 'feel' that there are use-cases but can't think of one atm. so maybe it is over-engineered.

File: nixos/modules/security/acme.nix (hunk 2 above)
Comment (qknight) [reply nixos/modules/security/acme.nix]: my main goal is `postfix.service` but maybe there are other use-cases as well.

**Comment by qknight** [comment]
ATM i wonder why there is: cert.domain at all. what is the motivation to not just use the attrName as 'domain' key? if no such motivation is there i'd like to rewrite the service and force the attrName to be the domain and get rid of the `domain` mkOption completely.

**Comment by fpletz** [comment]
The `domain` option was introduced to make it possible to have different certificates with the same common name (i.e. with different SubjAltNames for different services).

File: nixos/modules/security/acme.nix (hunk 2 above)
Comment (qknight) [reply nixos/modules/security/acme.nix]: we could use `dovecot.service` for instance

File: nixos/modules/security/acme.nix (hunk 3 above)
Comment (qknight) [reply nixos/modules/security/acme.nix]: many services as `postfix.service` and `dovecot.service` can load new certificates using `reload` so one does not have to cut connections, reset state and such which would be the case with a full reload.

but there could be a plugin for dovecot, a plugin like https://pigeonhole.dovecot.org/ for postifx, which hypotetically would require postfix to restart. i made this up of course, but the code for unique and restart over reload is already there and simple so why not just use it.

**Comment by qknight** [comment]
now that @fpletz explained why `domain` needs to be there i think that this patch is working for both of our use-cases.

**Comment by zimbatm** [comment]
@qknight I pushed a simplified proposal to your branch that takes into account what we where saying previously. It still needs a bit of testing. Feel free to revert or change it however you like.

**Comment by qknight** [comment]
@zimbatm happy to use your modified version. i have to find out what has to be done now, do i merge your stuff now and then sqash-merge it?

**Comment by zimbatm** [comment]
yes or keep adding commits to the branch and I'll squash them during merge

**Comment by qknight** [comment]
another thing i don't understand is `security.acme.certs.<name>.webroot`. in nixcloud-webservices i'm using a constant here, which points to `/var/lib/acme/acme-challenges` for all my `security.acme.certs`. 

i assume this value is closely coupled with the vhost abstraction in nginx you guys wrote, is that correct? i'd like to add a example value and a better description to it, any ideas what that could be?

**Comment by zimbatm** [comment]
yeah, I didn't build it but I'm pretty sure it's tied to the nginx enableTLS feature. It should work with Apache or any other webserver that can map the .well-known folder to that path. I believe that ACME has other validation methods, in which case this folder would be pretty useless.

**Comment by qknight** [comment]
@zimbatm if you like it, please merge it. i didn't test 'our' work in production yet, so merge with care. and thanks for your effort, very much appreciated!

**Comment by qknight** [comment]
@zimbatm @fpletz 

# regarding default values
at the moment this is the most important patch so that domain is always stet to a resonable value (not null but a string) by default. this is a important patch since i can read the certificates directly and don't have to check if they are null and otherwise use the attrName.

# regarding restart/reload
creating the new `reload` and `restart` targets was motivated by the idea that one can merge the different configs for a single domain so `nixcloud.email` could use the same TLS certificates as `nixcloud.webservices` but since merging does not work our current solution is to use seperate certs for these two use-cases.

this usage still does not exceed the quota of weekly certs of 4. that said, so far this works.

**Comment by qknight** [comment]
`webroot` might have to be a unique directory but i'm not sure.

**Comment by zimbatm** [comment]
How about creating a new PR with just the default values change?

The rest is less obvious to me and I don't have a setup to test it out right now; I think overall the idea is good, it just needs to be fleshed out a bit more.

**Comment by qknight** [comment]
@zimbatm yes, this seems to be a good idea. let's skip the restart/reload feature completely for the time being.

**Comment by qknight** [comment]
@zimbatm 

* please check this latest patch and squash merge it. 
* after your merge into master i would like to cherry-pick this into 17.09 later on, agreed?

**Comment by zimbatm** [comment]
@GrahamcOfBorg test acme

**Comment by zimbatm** [comment]
@qknight looks good to me. It should be safe to be cherry-picked as well.

Any idea on what to name the commit?

**Comment by qknight** [comment]
something like `acme_default_value_via_module_system`

**Review by dotlambda — COMMENTED** [review]
This breaks the build of the manual:
```
options-db.xml:6594: parser error : Opening and ending tag mismatch: name line 6594 and option
(<option>security.acme.cert.<name>.group</option>) to read SSL private certifica
```

**Comment by dotlambda** [comment]
Fixed XML in #34683

**Pull request merged**